libs/utils/Cargo.toml

```diff
@@ -31,10 +29,36 @@ serde_with.workspace = true
 once_cell.workspace = true
 strum.workspace = true
 strum_macros.workspace = true
-
+atty.workspace = true
 metrics.workspace = true
 pq_proto.workspace = true
 
+tracing.workspace = true
+
+# On every log event like `test_event` below
+# ```rust
+# let _entered = info_span!("timeline_detail", tenant = %tenant_id, timeline = %timeline_id).entered();
+# info!("test_event");
+# ```
+# there are static values: `info` log level, `timeline_detail` span name, etc.
+# and dynamic values: field names (`tenant`, `timeline`) and their values.
+#
+# Tracing allows the later to change during runtime, and generally cannot statically understand what's behind `%timeline_id` format call.
+# Due to that, env filters like `error,[{tenant=98d670ab7bee6f0051494306a1ab888f}]=debug` are split into two categories.
+# Static, `error` (another example could be `timeline_detail=error`) one that filters based on static event values,
+# and dynamic, `[{tenant=98d670ab7bee6f0051494306a1ab888f}]=debug`, that would require tracing to resolve the filter on every span
+# that might fit to the filter: some span with `tenant` field name and a dynamic value to resolve.
+#
+# Current tracing optimizes the span resolution, and filters out all `[{tenant=98d670ab7bee6f0051494306a1ab888f}]=debug` events because
+# there's a static filter with stricter log level: the global `error` one.
+# Patched version forces tracing to consider such dynamic filters as potential exceptions and check more potentially matching events
```

Comment:
what is the perf impact ?

Reply:
We could try using the label `run-benchmarks` pre-merge?

Reply:
It just occured to me that there probably isn't any codepaths exercised in the current benchmarks which would be have been changed, because there's fixed set of filters.

One "OLTP" benchmark would be:

```
pgbench -i -s 10  "hostname=127.0.0.1 port=55432 user=cloud_admin dbname=postgres"
pgbench -S -c 10 -T 100 -P 1  "hostname=127.0.0.1 port=55432 user=cloud_admin dbname=postgres"
pgbench -T 100 -c 10 -P 10  "hostname=127.0.0.1 port=55432 user=cloud_admin dbname=postgres"
pgbench -S -c 10 -T 100 -P 1  "hostname=127.0.0.1 port=55432 user=cloud_admin dbname=postgres"
```

(Used as "OLTP" with walredo work.)

Running this with `neon_local` on a clean `.neon` directory, then repeating it but with:
- another tenant (which doesn't participate in pgbench)
  - could have multiple timelines (branched at initdb_lsn it doesn't matter)
  - don't start pg for these

Then add for example more verbose logging for one timeline, confirm it's not logging much (for example, don't enable gc_loop to log everything), then silence other timeline.

This way the added filter paths would be exercised and we might see small degradation. But as long as we don't have hundreds of filters with matches I don't think there should be any effect...